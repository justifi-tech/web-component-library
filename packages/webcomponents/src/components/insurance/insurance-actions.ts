import { getErrorCode, getErrorMessage } from '../../api/services/utils';
import { ComponentErrorSeverity } from '../../api/ComponentError';

export const makeGetQuote = ({ authToken, service }) =>
  async ({ payload, onSuccess, onError }) => {
    try {
      const response = await service.fetchQuote(authToken, payload);

      if (!response.error) {
        const quote = response.data;
        onSuccess({ quote });
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        return onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      return onError({
        error: error.message || error,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };

export const makeToggleCoverage = ({ authToken, service }) =>
  async ({ quoteId, payload, onSuccess, onError }) => {
    try {
      const response = await service.toggleCoverage(authToken, quoteId, payload);

      if (!response.error) {
        const quote = response.data;
        onSuccess({ quote });
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        return onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      return onError({
        error: error.message || error,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };