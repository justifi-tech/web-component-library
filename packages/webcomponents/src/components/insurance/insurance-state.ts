import { createStore } from '@stencil/store';

const { state: insuranceValues, on: insuranceValuesOn } = createStore<any>({});

const { state: insuranceErrors } = createStore<any>({});

const validateInsuranceValues = () => {
  let invalid = false;

  Object.keys(insuranceValues).forEach((key) => {
    if (insuranceValues[key]) {
      insuranceErrors[key] = false;
    } else {
      insuranceErrors[key] = true;
      invalid = true;
    }
  });

  return { isValid: !invalid };
};

export { insuranceValues, insuranceValuesOn, insuranceErrors, validateInsuranceValues };
