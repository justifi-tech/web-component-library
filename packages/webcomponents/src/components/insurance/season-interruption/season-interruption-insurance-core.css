:host {
  display: block;
}

a {
  /*
    `!important` should be used sparingly if ever. In this case there is an inline
    style on links being returned from the server that needs to be overridden.
  */
  color: var(--jfi-body-color) !important;
}