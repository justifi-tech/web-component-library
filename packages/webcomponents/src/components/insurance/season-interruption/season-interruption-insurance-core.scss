@import 'reboot';
@import 'grid';

:host {
  @include host-base-styles;
  display: block;
}