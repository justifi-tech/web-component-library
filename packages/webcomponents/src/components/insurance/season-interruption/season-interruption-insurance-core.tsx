import { Component, h, Host, Prop, State, Event, EventEmitter } from "@stencil/core";
import { ComponentError } from "../../../api/ComponentError";
import { formatCurrency } from "../../../utils/utils";

const quoteResponse = {
  "id": "",
  "type": "array",
  "page_info": null,
  "data": {
    "id": "",
    "account_id": "",
    "accepted": false,
    "checkout_id": "",
    "currency": "USD",
    "expiration_date": "2024-05-31T00:00:00.000Z",
    "mode": "test",
    "policy_type": "season_interruption",
    "product": {
      "title": "Add Registration Protection",
      "description": "<p>Protect your registration and get your money back in the event of unforeseen circumstances that prevent participation, including sickness (including COVID-19), injury, inclement weather, transportation, job loss and other covered reasons.</p>",
      "legal_disclaimer": "<p>Selecting yes constitutes my electronic signature. I confirm that I have read, understand and agree to the <a href=\"https://partner.battleface.com/library/vertical-insure-season-cancellation-and-interruption\" target=\"_blank\" style=\"color: black\">policy terms and conditions</a>, and <a href=\"https://legal.verticalinsure.com/Vertical-Insure-Disclosures-SPI-Season-Can-and-Int.pdf\" target=\"_blank\" style=\"color: black\">important notices and disclosures</a>, which includes fraud warnings, privacy notice, and consent to electronic signature and delivery.</p>\n\n<p>battleface Travel Insurance plans are underwritten by Spinnaker Insurance Company (an IL Corporation, NAIC# 24376), with administrative office at One Pluckemin Way, Suite 102, Bedminster, NJ 07921. Plans are offered and administered by battleface Insurance Services LLC, 45 East Lincoln Street, Columbus, OH 43215, National Producer Number 18731960 (FL License number L107363/CA License number 0M75381).</p>"
    },
    "provider": "vertical_insure",
    "provider_id": "",
    "total_cents": 2000,
    "created_at": "2024-05-28T18:55:29.026Z",
    "updated_at": "2024-05-28T18:55:29.026Z"
  }
}

@Component({
  tag: 'justifi-season-interruption-insurance-core',
  styleUrl: 'season-interruption-insurance-core.scss',
  shadow: true,
})
export class SeasonInterruptionInsuranceCore {
  @Prop() getQuote: Function;

  @State() quote: any;
  @State() isLoading: boolean = false;

  @Event({ eventName: 'insurance-changed' }) insuranceChanged: EventEmitter;
  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    if (this.getQuote) {
      this.fetchData();
    }
  };

  fetchData(): void {
    this.isLoading = true;
    this.quote = quoteResponse.data;
    console.log(this.quote)
    // this.getQuote({
    //   onSuccess: ({ quote }) => {
    //     this.quote = quote
    //     this.isLoading = false;
    //   },
    //   onError: ({ error, code, severity }) => {
    //     this.isLoading = false;
    //     this.errorEvent.emit({
    //       errorCode: code,
    //       message: error,
    //       severity,
    //     });
    //   }
    // });
  };

  onChangeHandler(event: any) {
    console.log(event.target.value);
    this.insuranceChanged.emit();
    // await accept call
    // emit event to checkout
  }

  render() {
    return (
      <Host>
        <h2 class="fs-5 fw-bold pb-3 jfi-header">{this.quote.product.title}</h2>
        <div innerHTML={this.quote.product.description}></div>

        <div>
          <input
            id="accept"
            type="radio"
            name="opt-in"
            value="true"
            onChange={(event: any) => this.onChangeHandler(event)}
            class="btn-check jfi-btn-radio"
          />
          <label htmlFor="accept" class="btn btn-outline-primary jfi-btn-radio-label">
            Accept coverage for {formatCurrency(this.quote.total_cents)}
          </label>
        </div>
        <div>
          <input
            id="decline"
            type="radio"
            name="opt-in"
            value="false"
            onChange={(event: any) => this.onChangeHandler(event)}
            class="btn-check jfi-btn-radio"
          />
          <label htmlFor="decline" class="btn btn-outline-primary jfi-btn-radio-label">
            Decline coverage
          </label>
        </div>
        <div innerHTML={this.quote.product.legal_disclaimer}></div>
      </Host>
    );
  }
};