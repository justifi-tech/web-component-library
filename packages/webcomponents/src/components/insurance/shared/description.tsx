import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-description',
  shadow: true,
})

export class InsuranceDescription {
  render() {
    <div></div>
  };
};