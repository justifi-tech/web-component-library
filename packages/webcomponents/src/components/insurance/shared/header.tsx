import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-header',
  shadow: true,
})

export class InsuranceHeader {
  render() {
    <div></div>
  };
};