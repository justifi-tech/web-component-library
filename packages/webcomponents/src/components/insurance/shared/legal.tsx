import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-legal',
  shadow: true,
})

export class InsuranceLegal {
  render() {
    <div>
      <slot></slot>
    </div>
  };
};