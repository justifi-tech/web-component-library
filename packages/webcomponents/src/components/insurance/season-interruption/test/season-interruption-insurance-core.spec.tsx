jest.mock('../../../../ui-components/styled-host/styled-host.css', () => '');

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { SeasonInterruptionInsuranceCore } from '../season-interruption-insurance-core';
import { makeGetQuote } from '../../insurance-actions';
import { API_NOT_AUTHENTICATED_ERROR } from '../../../../api/shared';
import mockSeasonInterruptionInsurance from '../../../../../../../mockData/mockSeasonInterruptionInsurance.json';
import { insuranceValuesStore, insuranceErrorsStore } from '../../insurance-state';

describe('justifi-season-interruption-insurance-core', () => {
  beforeEach(() => insuranceValuesStore.dispose());
  beforeEach(() => insuranceErrorsStore.dispose());

  it('should display loading state correctly', async () => {
    const page = await newSpecPage({
      components: [SeasonInterruptionInsuranceCore],
      template: () => <justifi-season-interruption-insurance-core auth-token="123" />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('loads and sets the quote to state correctly', async () => {
    const getQuote = makeGetQuote({
      authToken: '123',
      service: {
        fetchQuote: jest.fn().mockResolvedValue(mockSeasonInterruptionInsurance),
      },
    })

    const page = await newSpecPage({
      components: [SeasonInterruptionInsuranceCore],
      template: () => <justifi-season-interruption-insurance-core auth-token="123" getQuote={getQuote} />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should emit error event when fetch fails', async () => {
    const getQuote = makeGetQuote({
      authToken: '',
      service: {
        fetchQuote: jest.fn().mockRejectedValue(new Error('Fetch error')),
      },
    })

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [SeasonInterruptionInsuranceCore],
      template: () => <justifi-season-interruption-insurance-core auth-token="123" getQuote={getQuote} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Fetch error',
          severity: 'error',
        }
      })
    );
  });

  it('should emit error event when API returns error', async () => {
    const getQuote = makeGetQuote({
      authToken: '',
      service: {
        fetchQuote: jest.fn().mockResolvedValue(API_NOT_AUTHENTICATED_ERROR),
      },
    })

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [SeasonInterruptionInsuranceCore],
      template: () => <justifi-season-interruption-insurance-core auth-token="123" getQuote={getQuote} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'not-authenticated',
          message: 'Not Authenticated',
          severity: 'error',
        }
      })
    );
  });

  it('validates that a selection was made', async () => {
    const getQuote = makeGetQuote({
      authToken: '123',
      service: {
        fetchQuote: jest.fn().mockResolvedValue(mockSeasonInterruptionInsurance),
      },
    })

    const page = await newSpecPage({
      components: [SeasonInterruptionInsuranceCore],
      template: () => <justifi-season-interruption-insurance-core auth-token="123" getQuote={getQuote} />,
    });

    await page.waitForChanges();

    const instance: any = page.rootInstance;
    const { isValid } = await instance.validate();

    await page.waitForChanges();

    expect(isValid).toBe(false); // Assuming provided fields pass the validation
    expect(page.root).toMatchSnapshot();
  });
});
