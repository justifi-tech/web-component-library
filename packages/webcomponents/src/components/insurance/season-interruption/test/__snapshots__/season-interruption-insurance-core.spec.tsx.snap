// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-season-interruption-insurance-core loads and sets the quote to state correctly 1`] = `
<justifi-season-interruption-insurance-core auth-token="123" class="body" style="display: block;">
  <style></style>
  <div>
    <h2 class="fs-5 fw-bold jfi-header pb-3">
      Add Registration Protection
    </h2>
    <small part="text">
      <p>
        Protect your registration and get your money back in the event of unforeseen circumstances that prevent participation, including sickness (including COVID-19), injury, inclement weather, transportation, job loss and other covered reasons.
      </p>
    </small>
    <div>
      <input class="form-check-input me-2" id="accept" name="opt-in" part="
          radio-input
          
          
        " type="radio" value="true">
      <label htmlfor="accept" part="radio-input-label">
        Accept coverage for $0.79
      </label>
    </div>
    <div class="mb-2">
      <input class="form-check-input me-2" id="decline" name="opt-in" part="
          radio-input
          
          
        " type="radio" value="false">
      <label htmlfor="decline" part="radio-input-label">
        Decline coverage
      </label>
    </div>
    <div class="invalid-feedback" style="display: none;">
      Please select an option
    </div>
    <small part="text">
      <p>
        Selecting yes constitutes my electronic signature. I confirm that I have read, understand and agree to the
        <a href="https://partner.battleface.com/library/vertical-insure-season-cancellation-and-interruption-es" target="_blank" style="color: black;">
          policy terms and conditions
        </a>
        , and
        <a href="https://legal.verticalinsure.com/Important-Notices-and-Disclosures-Everspan-Season-Can.pdf" target="_blank" style="color: black;">
          important notices and disclosures
        </a>
        , which includes fraud warnings, privacy notice, and consent to electronic signature and delivery.
      </p>
      <p>
        battleface Travel Insurance plans are underwritten by Everspan Insurance Company (an AZ Corporation, NAIC# 24961), with administrative office at One World Trade Center, 41st Floor, New York, NY 10007. Plans are offered and administered by battleface Insurance Services LLC, 45 East Lincoln Street, Columbus, OH 43215, National Producer Number 18731960 (FL License number L107363/CA License number 0M75381).
      </p>
    </small>
  </div>
</justifi-season-interruption-insurance-core>
`;

exports[`justifi-season-interruption-insurance-core should display loading state correctly 1`] = `
<justifi-season-interruption-insurance-core auth-token="123" class="body" style="display: block;">
  <style></style>
</justifi-season-interruption-insurance-core>
`;

exports[`justifi-season-interruption-insurance-core validates that a selection was made 1`] = `
<justifi-season-interruption-insurance-core auth-token="123" class="body" style="display: block;">
  <style></style>
  <div>
    <h2 class="fs-5 fw-bold jfi-header pb-3">
      Add Registration Protection
    </h2>
    <small part="text">
      <p>
        Protect your registration and get your money back in the event of unforeseen circumstances that prevent participation, including sickness (including COVID-19), injury, inclement weather, transportation, job loss and other covered reasons.
      </p>
    </small>
    <div>
      <input class="form-check-input is-invalid me-2" id="accept" name="opt-in" part="
          radio-input
          
          radio-input-invalid
        " type="radio" value="true">
      <label htmlfor="accept" part="radio-input-label">
        Accept coverage for $0.79
      </label>
    </div>
    <div class="mb-2">
      <input class="form-check-input is-invalid me-2" id="decline" name="opt-in" part="
          radio-input
          
          radio-input-invalid
        " type="radio" value="false">
      <label htmlfor="decline" part="radio-input-label">
        Decline coverage
      </label>
    </div>
    <div class="invalid-feedback" style="display: block;">
      Please select an option
    </div>
    <small part="text">
      <p>
        Selecting yes constitutes my electronic signature. I confirm that I have read, understand and agree to the
        <a href="https://partner.battleface.com/library/vertical-insure-season-cancellation-and-interruption-es" target="_blank" style="color: black;">
          policy terms and conditions
        </a>
        , and
        <a href="https://legal.verticalinsure.com/Important-Notices-and-Disclosures-Everspan-Season-Can.pdf" target="_blank" style="color: black;">
          important notices and disclosures
        </a>
        , which includes fraud warnings, privacy notice, and consent to electronic signature and delivery.
      </p>
      <p>
        battleface Travel Insurance plans are underwritten by Everspan Insurance Company (an AZ Corporation, NAIC# 24961), with administrative office at One World Trade Center, 41st Floor, New York, NY 10007. Plans are offered and administered by battleface Insurance Services LLC, 45 East Lincoln Street, Columbus, OH 43215, National Producer Number 18731960 (FL License number L107363/CA License number 0M75381).
      </p>
    </small>
  </div>
</justifi-season-interruption-insurance-core>
`;
